javascript/src/containers/Map.js

```diff
@@ -160,52 +161,77 @@ L.Icon.Default.mergeOptions({
     }
 
     updateCompositeLayer(urls) {
-        if ((urls.length === 0) && (this.compositeLayer !== null)) {
-            this.map.removeLayer(this.compositeLayer)
-            return
-        } else if (urls.length === 0) {
+        if (urls.length === 0) {
             return
         }
-
-        let tiles = new L.GridLayer()
-
-        tiles.createTile = function(coords) {
-            let tile = L.DomUtil.create('canvas', 'leaflet-tile')
-            let ctx = tile.getContext('2d')
-            let size = this.getTileSize()
-            tile.width = size.x
-            tile.height = size.y
-            let color = "green"
-            let loaded = 0
-            let images = []
-
-            urls.forEach((url) => {
-                let img = new Image()
-                img.onload = () => {
-                    loaded += 1
-                    if (loaded === urls.length){
-                        images.forEach((image, i) => {
-                            ctx.globalCompositeOperation = (i === 0 ? "source-out" : "source-in")
-                            ctx.drawImage(image, 0, 0)
-                        })
-                        if (color) {
-                            ctx.globalCompositeOperation = "source-atop" // color existing pixels
-                            ctx.fillStyle = color
-                            ctx.fillRect(0, 0, tile.width, tile.height)
-                        }
-                    }
+        let DivLayer = L.GridLayer.extend({
+            createTile: function (coords) {
+                let tile = L.DomUtil.create('div', 'leaflet-tile')
+                let size = this.getTileSize()
+                let canvases = {
+                    multiHab: null,
+                    lostHab: null,
+                    keptHab: null,
+                    newHab: null
                 }
-                img.src = `${url}/${coords.z}/${coords.x}/${coords.y}.png`
-                images.push(img)
-            })
 
-            return tile
-        }
+                Object.keys(canvases).forEach((canvas) => {
+                    canvases[canvas] = L.DomUtil.create('canvas', 'habitat-canvas', tile)
+                    canvases[canvas].width = size.x
+                    canvases[canvas].height = size.y
+                })
 
-        if (this.compositeLayer !== null) {
+                let color = "green"
+                let loaded = 0
+                let images = []
+                let drawHabitatCanvas = (canvas, m1, m2, op, color) => {
+
+                    let ctx = canvas.getContext('2d');
+                    ctx.drawImage(m1, 0, 0)
+                    ctx.globalCompositeOperation = op
+                    ctx.drawImage(m2, 0, 0)
+                    // add color
+                    ctx.globalCompositeOperation = "source-atop"
+                    ctx.fillStyle = color
+                    ctx.fillRect(0, 0, canvas.width, canvas.height)
+                }
+
+                urls.forEach((url) => {
```

Comment:
Here's an (untested) idea that is a bit more verbose, but cleaner than counting # of loads. Create a promise around image loading then use `Promise.all()` to wait for all loads to complete:

```javascript
let promises = urls.map(url => {
    return Promise(resolve => {
        let img = new Image()
        img.onload = () => resolve(img)
        img.src = `${url}/${coords.z}/${coords.x}/${coords.y}.png`
    })
})
Promise.all(promises).then(images => {
    if (urls.length !== 2) {
        // ...
    }
})
```

Also returns the list of images in the `Promise.all()` callback so no need to keep an array of them.